src/sgame/sg_bot.cpp

```diff
@@ -263,6 +262,188 @@ bool G_BotSetDefaults( int clientNum, team_t team, int skill, Str::StringRef beh
+};
+
+static const std::vector<botSkillTreeElement_t> fighting_skills = {
+	// aliens
+	{ "aim-head",           BOT_A_AIM_HEAD,      10, pred_alien, {} },
+
+	// humans
+	{ "predict-aim",        BOT_H_PREDICTIVE_AIM, 5, pred_human, {} },
+};
+
+static const std::vector<botSkillTreeElement_t> initial_unlockable_skills = {
+	// movement skills
+	{ "movement", BOT_B_BASIC_MOVEMENT, 2, pred_always, movement_skills },
+	// fighting skills
+	{ "fighting", BOT_B_BASIC_FIGHT, 2, pred_always, fighting_skills },
+	// situation awareness and survival
+	{ "feels-pain", BOT_B_PAIN, 2, pred_always, survival_skills },
+};
+
+// Note: this function modifies possible_choices to remove the one we just
+//       chose, and add the new unlocked skills to that list.
+static Util::optional<botSkillTreeElement_t> ChooseOneSkill(const gentity_t *bot, skillSet_t skillSet, std::vector<botSkillTreeElement_t> &possible_choices, std::mt19937_64& rng)
+{
+	int total_skill_points = 0;
+	for (const auto &skill : possible_choices)
+	{
+		if (skill.predicate(bot, skillSet))
+		{
+			total_skill_points += skill.cost;
+		}
+	}
+
+	if (total_skill_points <= 0)
+		return Util::nullopt;
+
+	std::uniform_int_distribution<int> dist(0, total_skill_points - 1);
+	int random = dist(rng);
+
+	// find which skill this corresponds to (higher point skills are more
+	// likely to be found)
+	int cursor = 0;
+	for (std::vector<botSkillTreeElement_t>::iterator skill = possible_choices.begin(); skill != possible_choices.end(); skill++)
+	{
+		if (!skill->predicate(bot, skillSet))
+		{
+			continue;
+		}
+
+		cursor += skill->cost;
+		if (random < cursor)
+		{
+			botSkillTreeElement_t result = *skill;
+
+			// update the list of future choices
+			possible_choices.erase(skill);
+			for (const auto &child : result.unlocked_skills)
+			{
+				possible_choices.push_back(std::move(child));
+			}
+			result.unlocked_skills.clear();
+
+			return result;
+		}
+	}
+	return Util::nullopt;
+}
+
+// This selects a set of skills for a bot.
+//
+// The algorithm is stable and deterministic.
+// Deterministic:
+//     a bot's skill set is determined only by the bot's name and the bot's
+//     skill level. A different bot will have a different skillset, but this
+//     means that a bot with the same name will have the same skillset from
+//     a game to another.
+// Stable:
+//     when a bot changes skill level, the first skills unlocked are common
+//     between both skill levels. This means that a bot whose skill is raised
+//     by BotSetSkillLevel will keep the skills it already had, and get new
+//     ones.
+//     This essentially means that a bot will keep its "personality" after
+//     a skill change, although keeping its base skills or unlocking new ones.
+std::pair<std::string, skillSet_t> BotDetermineSkills(gentity_t *bot, int skill)
+{
+	std::vector<botSkillTreeElement_t> possible_choices = initial_unlockable_skills;
+	// aliens have 50 points to spend max
```

Comment:
What if instead of hardcoding spendable points, we made skill be a percentage of the max possible skills people could buy. For example,

if the total cost of all human skill is 100, then skill level 9 would mean that you would have 90 points to spend?

Reply:
I initially wanted to do that, but there are two problems to do that:

First, the `botSkillTreeElement_t->predicate` functions makes this problem undecidable in the general sense: you can't count how many skill point you might spend at most because you could have skills that are excluding each other (for example some aggressive skill could be disabled once you unlock a defensive skill using a `bool pred_x_not_unlocked(const gentity_t *, skillSet_t existing_skills) { return existing_skills[BOT_A_XYZ]; }`).

A second problem is a balancing one. I've found that doing "skill 5 has 5/9*max skill points" actually nerfs human bots a bit too much in my opinion. I manually tweaked the value up a bit to compensate for that. I don't think it's a bad idea to keep that freedom.